File: JavaApp/src/main/java/com/charlezz/javaapp/feature/remote/PostActivityModule.java
package com.charlezz.javaapp.feature.remote;


import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;

import com.charlezz.javaapp.databinding.ActivityPostBinding;
import com.charlezz.javaapp.di.ActivityScope;
import com.charlezz.javaapp.di.FragmentScope;

import dagger.Module;
import dagger.Provides;
import dagger.android.ContributesAndroidInjector;

@Module
public abstract class PostActivityModule {
    @Provides
    @ActivityScope
    static ActivityPostBinding provideBiding(PostActivity activity){
        return DataBindingUtil.setContentView(activity, R.layout.activity_post);
    }

    @Provides
    @ActivityScope
    static PostViewModel provideViewModel(PostActivity activity){
        return ViewModelProviders.of(activity).get(PostViewModel.class);
    }

    @FragmentScope
    @ContributesAndroidInjector(modules = {PostFragment.class})
    abstract PostFragment postFragment();
}

File: app/src/main/java/com/charlezz/finalarchitecture/di/module/remote/PostActivityModule.kt
package com.charlezz.finalarchitecture.di.module.remote

import android.databinding.DataBindingUtil
import com.charlezz.finalarchitecture.R
import com.charlezz.finalarchitecture.databinding.ActivityPostBinding
import com.charlezz.finalarchitecture.di.annotation.ActivityScope
import com.charlezz.finalarchitecture.di.annotation.FragmentScope
import com.charlezz.finalarchitecture.ui.remote.PostActivity
import com.charlezz.finalarchitecture.ui.remote.PostFragment
import dagger.Module
import dagger.Provides
import dagger.android.ContributesAndroidInjector

@Module
abstract class PostActivityModule{
    @Module
    companion object {
        @JvmStatic
        @Provides
        @ActivityScope
        fun provideActivityPostBinding(activity: PostActivity): ActivityPostBinding =
                DataBindingUtil.setContentView(activity, R.layout.activity_post)
    }

    @FragmentScope
    @ContributesAndroidInjector(modules = [(PostFragmentModule::class)])
    abstract fun getPostFragment(): PostFragment

}